ampd/src/event_processor.rs

```diff
@@ -63,85 +67,82 @@ impl Default for Config {
 /// Let the `handler` consume events from the `event_stream`. The token is checked for cancellation
 /// at the end of each consumed block or when the `event_stream` times out. If the token is cancelled or the
 /// `event_stream` is closed, the function returns
-pub async fn consume_events<H, B, S, E>(
+#[instrument(fields(handler = %handler_label), skip_all)]
+pub async fn consume_events<H, S, C>(
     handler_label: String,
     handler: H,
-    broadcaster: B,
     event_stream: S,
     event_processor_config: Config,
     token: CancellationToken,
+    msg_queue_client: broadcaster_v2::MsgQueueClient<C>,
     metric_client: monitoring::Client,
 ) -> Result<(), Error>
 where
     H: EventHandler,
-    B: BroadcasterClient,
-    S: Stream<Item = Result<Event, E>>,
-    E: Context,
+    S: Stream<Item = Result<Event, event_sub::Error>>,
+    C: cosmos::CosmosClient + Clone,
 {
-    let mut event_stream = Box::pin(event_stream);
-    loop {
-        let stream_status =
-            retrieve_next_event(&mut event_stream, event_processor_config.stream_timeout)
-                .await
-                .change_context(Error::EventStream)?;
-
-        if let StreamStatus::Active(event) = &stream_status {
-            handle_event(
-                &handler,
-                &broadcaster,
-                event,
-                RetryPolicy::RepeatConstant {
-                    sleep: event_processor_config.retry_delay,
-                    max_attempts: event_processor_config.retry_max_attempts,
-                },
-            )
-            .await?;
-        }
-
-        if let StreamStatus::Active(Event::BlockEnd(height)) = &stream_status {
-            info!(
-                handler = handler_label,
-                height = height.value(),
-                "handler finished processing block"
-            );
-
-            if let Err(err) = metric_client.metrics().record_metric(Msg::IncBlockReceived) {
-                warn!( handler = handler_label,
-                    height = height.value(),
-                    err = %err,
-                    "failed to record block received metric",
-                );
+    let Config {
+        retry_delay,
+        retry_max_attempts,
+        stream_timeout,
+        ..
+    } = event_processor_config;
+    let handler_retry = RetryPolicy::repeat_constant(retry_delay, retry_max_attempts);
+
+    let event_stream =
+        tokio_stream::StreamExt::timeout_repeating(event_stream, time::interval(stream_timeout));
```

Comment:
did you make sure this doesn't duplicate capabilities the event stream already has?

Reply:
I don't see any timeout implementation on the underlying stream if that's what you mean